File: src/cli/collection/schema.ts
import { Command, ExactArgs, ref } from "vipvot";
import { printJson, printError } from "../../lib/output.ts";
import { getSettings } from "../../lib/config.ts";
import { getMongoClient, closeAllClients } from "../../mongo/client.ts";
import { inferSchema } from "../../mongo/schema.ts";
import type { SchemaResult } from "../../mongo/schema.ts";
import { resolveConnectionAlias } from "../_globals.ts";

export function buildSchemaCommand(): Command {
  const sampleSize = ref<string>("");
  const depth = ref<string>("");
  const limit = ref<string>("");
  const skip = ref<string>("");

  const cmd = Command({
    use: "schema <database> <collection>",
    short: "Infer collection schema by sampling documents",
    args: ExactArgs(2),
    run: async (_cmd, args) => {
      const [database, collection] = args as [string, string];
      try {
        const defaultSize = getSettings().defaults?.schemaSampleSize ?? 100;
        const sample = sampleSize.value ? parseInt(sampleSize.value, 10) : defaultSize;
        if (!Number.isFinite(sample) || sample < 1) {
          throw new Error(
            `Invalid --sample-size: "${sampleSize.value}". Must be a positive integer.`,
          );
        }

        const maxDepth = parsePositiveInt(depth.value || undefined, "--depth");
        const limitVal = parsePositiveInt(limit.value || undefined, "--limit");
        const skipVal = skip.value ? parseNonNegativeInt(skip.value, "--skip") : 0;

        const { client } = await getMongoClient(resolveConnectionAlias());
        const result = await inferSchema(client, {
          dbName: database,
          collName: collection,
          sampleSize: sample,
          maxDepth,
        });

        printSchemaResult(result, { limit: limitVal, skip: skipVal });
      } catch (err) {
        printError(err instanceof Error ? err.message : "Failed to infer schema");
      } finally {
        await closeAllClients();
      }
    },
  });

  cmd.flags().stringVarP(sampleSize, "sample-size", "", "", "Number of documents to sample");
  cmd
    .flags()
    .stringVarP(depth, "depth", "", "", "Max nesting depth for fields (1 = top-level only)");
  cmd.flags().stringVarP(limit, "limit", "", "", "Max fields to return (for pagination)");
  cmd.flags().stringVarP(skip, "skip", "", "", "Number of fields to skip (for pagination)");

  return cmd;
}

function printSchemaResult(
  result: SchemaResult,
  { limit, skip }: { limit: number | undefined; skip: number },
): void {
  const totalFields = result.fields.length;
  const slicedFields =
    limit !== undefined ? result.fields.slice(skip, skip + limit) : result.fields.slice(skip);
  const hasMore = skip + slicedFields.length < totalFields;

  const output: Record<string, unknown> = {
    database: result.database,
    collection: result.collection,
    sampleSize: result.sampleSize,
    totalDocuments: result.totalDocuments,
    totalFields,
    fields: slicedFields,
  };

  if (hasMore) {
    output.pagination = {
      hasMore: true,
      nextSkip: skip + slicedFields.length,
    };
  }

  printJson(output);
}

function parsePositiveInt(value: string | undefined, name: string): number | undefined {
  if (value === undefined) {
    return undefined;
  }
  const num = parseInt(value, 10);
  if (!Number.isFinite(num) || num < 1) {
    throw new Error(`Invalid ${name}: "${value}". Must be a positive integer.`);
  }
  return num;
}

function parseNonNegativeInt(value: string, name: string): number {
  const num = parseInt(value, 10);
  if (!Number.isFinite(num) || num < 0) {
    throw new Error(`Invalid ${name}: "${value}". Must be a non-negative integer.`);
  }
  return num;
}

File: internal/cli/collection/schema.go
package collection

import (
	"maps"
	"strconv"

	"github.com/spf13/cobra"

	"github.com/shhac/agent-mongo/internal/cli/shared"
	"github.com/shhac/agent-mongo/internal/config"
	"github.com/shhac/agent-mongo/internal/mongo"
	"github.com/shhac/agent-mongo/internal/output"
)

func registerSchema(parent *cobra.Command, globals func() *shared.GlobalFlags) {
	var sampleSize, depth, limit, skip string

	cmd := &cobra.Command{
		Use:   "schema <database> <collection>",
		Short: "Infer collection schema by sampling documents",
		Args:  cobra.ExactArgs(2),
		RunE: func(_ *cobra.Command, args []string) error {
			g := globals()
			ref := mongo.Ref{DB: args[0], Collection: args[1]}

			sample, err := shared.ParsePositiveInt(sampleSize, "--sample-size")
			if err != nil {
				return err
			}
			if sample == 0 {
				sample = config.SettingOr("defaults.schemaSampleSize")
			}
			maxDepth, err := shared.ParsePositiveInt(depth, "--depth")
			if err != nil {
				return err
			}
			limitVal, err := shared.ParsePositiveInt(limit, "--limit")
			if err != nil {
				return err
			}
			skipVal, err := shared.ParseNonNegativeInt(skip, "--skip")
			if err != nil {
				return err
			}

			return shared.WithSessionRef(g, ref, func(ctx shared.SessionCtx) error {
				result, err := ctx.Session.InferSchema(ctx.Ctx, mongo.SchemaOpts{
					Ref:        ref,
					SampleSize: sample,
					MaxDepth:   maxDepth,
				})
				if err != nil {
					return err
				}
				return printSchema(result, limitVal, skipVal)
			})
		},
	}

	cmd.Flags().StringVar(&sampleSize, "sample-size", "", "Number of documents to sample")
	cmd.Flags().StringVar(&depth, "depth", "", "Max nesting depth for fields (1 = top-level only)")
	cmd.Flags().StringVar(&limit, "limit", "", "Max fields to return (for pagination)")
	cmd.Flags().StringVar(&skip, "skip", "", "Number of fields to skip (for pagination)")
	parent.AddCommand(cmd)
}

// paginateFields slices the field list for --limit/--skip pagination.
// limit == 0 means unlimited.
func paginateFields(fields []mongo.FieldInfo, skip, limit int) ([]mongo.FieldInfo, bool) {
	if skip >= len(fields) {
		return nil, false
	}
	sliced := fields[skip:]
	if limit > 0 && len(sliced) > limit {
		sliced = sliced[:limit]
	}
	hasMore := skip+len(sliced) < len(fields)
	return sliced, hasMore
}

func printSchema(result mongo.SchemaResult, limit, skip int) error {
	sliced, hasMore := paginateFields(result.Fields, skip, limit)

	meta := output.Meta(map[string]any{
		"database":       result.Database,
		"collection":     result.Collection,
		"sampleSize":     result.SampleSize,
		"totalDocuments": result.TotalDocuments,
		"totalFields":    len(result.Fields),
	})
	if hasMore {
		nextSkip := strconv.Itoa(skip + len(sliced))
		maps.Copy(meta, output.PaginationMeta(true, nextSkip, len(result.Fields)))
	}
	return output.PrintList(sliced, meta)
}
